actors/miner/src/vesting_state.rs

```diff
@@ -1,54 +1,112 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::{iter, mem};
+use std::iter;
 
+use cid::Cid;
+use fil_actors_runtime::{ActorError, AsActorError};
+use fvm_ipld_blockstore::Blockstore;
 use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::CborStore;
 use fvm_shared::clock::ChainEpoch;
 use fvm_shared::econ::TokenAmount;
+use fvm_shared::error::ExitCode;
 use itertools::{EitherOrBoth, Itertools};
+use multihash_codetable::Code;
 use num_traits::Zero;
 
 use super::{QuantSpec, VestSpec};
 
 // Represents miner funds that will vest at the given epoch.
-#[derive(Debug, Serialize_tuple, Deserialize_tuple)]
+#[derive(Default, Debug, Serialize_tuple, Deserialize_tuple, Clone)]
 pub struct VestingFund {
     pub epoch: ChainEpoch,
     pub amount: TokenAmount,
 }
 
-/// Represents the vesting table state for the miner.
-/// It is a slice of (VestingEpoch, VestingAmount).
-/// The slice will always be sorted by the VestingEpoch.
-#[derive(Serialize_tuple, Deserialize_tuple, Default)]
+/// Represents the vesting table state for the miner. It's composed of a `head` (stored inline) and
+/// a tail (referenced by CID).
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
 pub struct VestingFunds {
-    pub funds: Vec<VestingFund>,
+    // The "next" batch of vesting funds:
+    // - If this is None, there are no vesting funds.
+    // - All batches in `tail` are guaranteed to vest after this batch.
+    // - This batch _can_ be empty (have a zero amount) if we've burnt through it (fees &
+    //   penalties).
+    head: Option<VestingFund>,
+    // The rest of the vesting funds, if any.
+    tail: Cid, // Vec<VestingFund>
 }
 
 impl VestingFunds {
-    pub fn new() -> Self {
-        Default::default()
+    pub fn new(store: &impl Blockstore) -> Result<Self, ActorError> {
+        let tail = store
+            .put_cbor(&Vec::<VestingFund>::new(), Code::Blake2b256)
+            .context_code(ExitCode::USR_ILLEGAL_STATE, "failed to construct vesting funds")?;
+        Ok(Self { head: None, tail })
     }
 
-    pub fn unlock_vested_funds(&mut self, current_epoch: ChainEpoch) -> TokenAmount {
-        // TODO: the funds are sorted by epoch, so we could do a binary search here
-        let i = self
-            .funds
-            .iter()
-            .position(|fund| fund.epoch >= current_epoch)
-            .unwrap_or(self.funds.len());
+    pub fn load(
+        &self,
+        store: &impl Blockstore,
+    ) -> Result<impl Iterator<Item = VestingFund>, ActorError> {
+        // NOTE: we allow head to be drawn down to zero through fees/penalties. However, when
+        // inspecting the vesting table, we never want to see a "zero" entry so we skip it in that case.
+        // We don't set it to "none" in that case because "none" means that we have _no_ vesting funds.
```

Comment:
Well maybe this argues for `vesting_funds` to be optional in the miner actor? maybe that's a better signal for no-funds than having a zero `VestingFund` at `head`?

Reply:
done. Doing it in `VestingFunds` was pretty straightforward.